core/src/test/java/org/elasticsearch/index/seqno/LocalCheckpointServiceTests.java

```diff
@@ -0,0 +1,192 @@
+
+        for (int i = 0; i < maxOps; i++) {
+            seqNoList.add(i);
+        }
+        Collections.shuffle(seqNoList, random());
+        for (Integer seqNo : seqNoList) {
+            checkpointService.markSeqNoAsCompleted(seqNo);
+        }
+        assertThat(checkpointService.checkpoint, equalTo(maxOps - 1L));
+        assertThat(checkpointService.processedSeqNo.size(), equalTo(aligned ? 0 : 1));
+    }
+
+    public void testConcurrentPrimary() throws InterruptedException {
+        Thread[] threads = new Thread[randomIntBetween(2, 5)];
+        final int opsPerThread = randomIntBetween(10, 20);
+        final int maxOps = opsPerThread * threads.length;
+        final long unFinishedSeq = randomIntBetween(0, maxOps - 2); // make sure we always index the last seqNo to simplify maxSeq checks
+        logger.info("--> will run [{}] threads, maxOps [{}], unfinished seq no [{}]", threads.length, maxOps, unFinishedSeq);
+        final CyclicBarrier barrier = new CyclicBarrier(threads.length);
+        for (int t = 0; t < threads.length; t++) {
+            final int threadId = t;
+            threads[t] = new Thread(new AbstractRunnable() {
+                @Override
+                public void onFailure(Throwable t) {
+                    throw new ElasticsearchException("failure in background thread", t);
+                }
+
+                @Override
+                protected void doRun() throws Exception {
+                    barrier.await();
+                    for (int i = 0; i < opsPerThread; i++) {
+                        long seqNo = checkpointService.generateSeqNo();
+                        logger.info("[t{}] started   [{}]", threadId, seqNo);
+                        if (seqNo != unFinishedSeq) {
+                            checkpointService.markSeqNoAsCompleted(seqNo);
+                            logger.info("[t{}] completed [{}]", threadId, seqNo);
+                        }
+                    }
+                }
+            }, "testConcurrentPrimary_" + threadId);
+            threads[t].start();
+        }
+        for (Thread thread : threads) {
+            thread.join();
+        }
+        assertThat(checkpointService.getMaxSeqNo(), equalTo(maxOps - 1L));
+        assertThat(checkpointService.getCheckpoint(), equalTo(unFinishedSeq - 1L));
+        checkpointService.markSeqNoAsCompleted(unFinishedSeq);
+        assertThat(checkpointService.getCheckpoint(), equalTo(maxOps - 1L));
+        assertThat(checkpointService.processedSeqNo.size(), isOneOf(0, 1));
+    }
+
+    public void testConcurrentReplica() throws InterruptedException {
+        Thread[] threads = new Thread[randomIntBetween(2, 5)];
+        final int opsPerThread = randomIntBetween(10, 20);
+        final int maxOps = opsPerThread * threads.length;
+        final long unFinishedSeq = randomIntBetween(0, maxOps - 2); // make sure we always index the last seqNo to simplify maxSeq checks
+        Set<Integer> seqNoList = new HashSet<>();
```

Comment:
`Set<Integer> seqNoList = IntStream.range(0, maxOps).boxed().collect(Collectors.toSet());` :) and maybe change the name to `seqNos` (at least remove `List` from the name)?